wrappers/tencentcloudkms/tencentcloudkms.go

```diff
@@ -0,0 +1,252 @@
+package tencentcloudkms
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"sync/atomic"
+
+	wrapping "github.com/hashicorp/go-kms-wrapping"
+	"github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/common"
+	"github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/common/profile"
+	kms "github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/kms/v20190118"
+)
+
+// These constants contain the accepted env vars; the Vault one is for backwards compat
+const (
+	EnvTencentCloudKMSWrapperKeyID   = "TENCENTCLOUDKMS_WRAPPER_KEY_ID"
+	EnvVaultTencentCloudKMSSealKeyID = "VAULT_TENCENTCLOUDKMS_SEAL_KEY_ID"
+)
+
+// These constants are TencentCloud accepted env vars
+const (
+	PROVIDER_SECRET_ID      = "TENCENTCLOUD_SECRET_ID"
+	PROVIDER_SECRET_KEY     = "TENCENTCLOUD_SECRET_KEY"
+	PROVIDER_SECURITY_TOKEN = "TENCENTCLOUD_SECURITY_TOKEN"
+	PROVIDER_REGION         = "TENCENTCLOUD_REGION"
+)
+
+// Wrapper is a wrapper that uses TencentCloud KMS
+type Wrapper struct {
+	accessKey    string
+	secretKey    string
+	sessionToken string
+	region       string
+
+	keyID        string
+	currentKeyID *atomic.Value
```

Comment:
I'm not sure about this currentKeyID mechanism. If this value can be changed on every encrypt operation, then it seems like any time you do a KeyID() call, this value might have been overwritten multiple times. In other words, I'm not sure it's meaningful to display the current actual value as opposed to the original alias.

Reply:
Hello @jefferai , thanks for your reviews.

For tencentcloud kms, every encrypt operation will returns a unique `encrypt key id`, which I stores it to `currentKeyID`, this `encrypt key id` is totally different from the `kms key id` that setting in the configure. When do a `KeyID()` call, it just returns the `encrypt key id`, it is a read function.

Reply:
That's my point though. The KeyID call will return different values over time, but that is not necessarily the value you're expecting. Let's say you have multiple goroutines using this wrapper. Routine A does an encrypt, then B does an encrypt. Then A reads KeyID but what it gets back isn't the key ID that corresponded to the encrypt call it performed, but instead it gets the key ID corresponding to B's encryption call.

Reply:
Are you sure _every_ encrypt operation returns a _unique_ key ID? As opposed to the key changing over time due to e.g. periodic rotation?

Reply:
Yes, every encrypt operation uses a golbal `kms key id` and a unique `encrypt key id`. You may image that every encrypt operation uses a key (`kms key id`) and a salt (`encrypt key id`).

Reply:
> That's my point though. The KeyID call will return different values over time, but that is not necessarily the value you're expecting. Let's say you have multiple goroutines using this wrapper. Routine A does an encrypt, then B does an encrypt. Then A reads KeyID but what it gets back isn't the key ID that corresponded to the encrypt call it performed, but instead it gets the key ID corresponding to B's encryption call.

So, what is your suggestions? remove the function `KeyID()` ?

Reply:
> Are you sure _every_ encrypt operation returns a _unique_ key ID? As opposed to the key changing over time due to e.g. periodic rotation?

Sorry, I correct what I said: tencentcloud always returns the fixed `kms key id`, what I said below is wrong.